App crash (pick image/images from GDrive without Crop option)

App crash after pick images from Google Drive (or any remote storage) without cropping
Logs:
`  java.lang.RuntimeException: Unable to destroy activity {com.fuck_boilerplate.rx_paparazzo.sample/rx_activity_result.HolderActivity}: rx.exceptions.OnErrorNotImplementedException: Attempt to invoke virtual method 'char[] java.lang.String.toCharArray()' on a null object reference
                                                                                              at android.app.ActivityThread.performDestroyActivity(ActivityThread.java:3831)
                                                                                              at android.app.ActivityThread.handleDestroyActivity(ActivityThread.java:3849)
                                                                                              at android.app.ActivityThread.-wrap5(ActivityThread.java)
                                                                                              at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1398)
                                                                                              at android.os.Handler.dispatchMessage(Handler.java:102)
                                                                                              at android.os.Looper.loop(Looper.java:148)
                                                                                              at android.app.ActivityThread.main(ActivityThread.java:5417)
                                                                                              at java.lang.reflect.Method.invoke(Native Method)
                                                                                              at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:726)
                                                                                              at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:616)
                                                                                           Caused by: rx.exceptions.OnErrorNotImplementedException: Attempt to invoke virtual method 'char[] java.lang.String.toCharArray()' on a null object reference
                                                                                              at rx.internal.util.InternalObservableUtils$ErrorNotImplementedAction.call(InternalObservableUtils.java:374)
                                                                                              at rx.internal.util.InternalObservableUtils$ErrorNotImplementedAction.call(InternalObservableUtils.java:371)
                                                                                              at rx.internal.util.ActionSubscriber.onError(ActionSubscriber.java:44)
                                                                                              at rx.observers.SafeSubscriber._onError(SafeSubscriber.java:157)
                                                                                              at rx.observers.SafeSubscriber.onError(SafeSubscriber.java:120)
                                                                                              at rx.exceptions.Exceptions.throwOrReport(Exceptions.java:204)
                                                                                              at rx.internal.operators.OperatorOnErrorResumeNextViaFunction$4.onError(OperatorOnErrorResumeNextViaFunction.java:143)
                                                                                              at rx.internal.operators.OperatorMap$MapSubscriber.onError(OperatorMap.java:85)
                                                                                              at rx.internal.operators.OperatorMerge$MergeSubscriber.reportError(OperatorMerge.java:266)
                                                                                              at rx.internal.operators.OperatorMerge$MergeSubscriber.checkTerminate(OperatorMerge.java:810)
                                                                                              at rx.internal.operators.OperatorMerge$MergeSubscriber.emitLoop(OperatorMerge.java:571)
                                                                                              at rx.internal.operators.OperatorMerge$MergeSubscriber.emit(OperatorMerge.java:560)
                                                                                              at rx.internal.operators.OperatorMerge$InnerSubscriber.onError(OperatorMerge.java:844)
                                                                                              at rx.observers.Subscribers$5.onError(Subscribers.java:224)
                                                                                              at rx.exceptions.Exceptions.throwOrReport(Exceptions.java:192)
                                                                                              at rx.internal.operators.OperatorZip$Zip.tick(OperatorZip.java:269)
                                                                                              at rx.internal.operators.OperatorZip$Zip$InnerSubscriber.onNext(OperatorZip.java:335)
                                                                                              at rx.internal.operators.OperatorMap$MapSubscriber.onNext(OperatorMap.java:74)
                                                                                              at rx.internal.util.ScalarSynchronousObservable$WeakSingleProducer.request(ScalarSynchronousObservable.java:268)
                                                                                              at rx.Subscriber.setProducer(Subscriber.java:211)
                                                                                              at rx.internal.operators.OperatorMap$MapSubscriber.setProducer(OperatorMap.java:99)
                                                                                              at rx.internal.util.ScalarSynchronousObservable$1.call(ScalarSynchronousObservable.java:79)
                                                                                              at rx.internal.util.ScalarSynchronousObservable$1.call(ScalarSynchronousObservable.java:75)
                                                                                              at rx.internal.operators.OnSubscribeLift.call(OnSubscribeLift.java:50)
                                                                                              at rx.internal.operators.OnSubscribeLift.call(OnSubscribeLift.java:30)
                                                                                              at rx.Observable.unsafeSubscribe(Observable.java:8460)
                                                                                              at rx.internal.operators.OperatorZip$Zip.start(OperatorZip.java:214)
                                                                                              at rx.internal.operators.OperatorZip$ZipSubscriber.onNext(OperatorZip.java:156)
                                                                                              at rx.internal.operators.OperatorZip$ZipSubscriber.onNext(OperatorZip.java:122)
                                                                                              at rx.internal.util.ScalarSynchronousObservable$WeakSingleProducer.request(ScalarSynchronousObservable.java:268)
                                                                                              at rx.Subscriber.setProducer(Subscriber.java:209)
                                                                                              at rx.internal.util.ScalarSynchronousObservable$1.call(ScalarSynchronousObservable.java:79)
                                                                                              at rx.internal.util.ScalarSynchronousObservable$1.call(ScalarSynchronousObservable.java:75)
                                                                                              at rx.internal.operators.OnSubscribeLift.call(OnSubscribeLift.java:50)
                                                                                              at rx.internal.operators.OnSubscribeLift.call(OnSubscribeLift.java:30)
                                                                                          	at rx.Observable.unsafe
`

--- a/app/src/androidTest/java/com/fuck_boilerplate/rx_paparazzo/sample/ApplicationTest.java
+++ b/app/src/androidTest/java/com/fuck_boilerplate/rx_paparazzo/sample/ApplicationTest.java
@@ -21,8 +21,10 @@ import android.widget.ImageView;
 import com.fuck_boilerplate.rx_paparazzo.entities.Config;
 import com.fuck_boilerplate.rx_paparazzo.entities.Size;
 import com.fuck_boilerplate.rx_paparazzo.entities.TargetUi;
+import com.fuck_boilerplate.rx_paparazzo.interactors.DownloadImage;
 import com.fuck_boilerplate.rx_paparazzo.interactors.GetDimens;
 import com.fuck_boilerplate.rx_paparazzo.interactors.GetPath;
+import com.fuck_boilerplate.rx_paparazzo.interactors.ImageUtils;
 import com.fuck_boilerplate.rx_paparazzo.sample.activities.StartActivity;
 import com.fuck_boilerplate.rx_paparazzo.sample.activities.Testable;
 
@@ -215,10 +217,12 @@ public class ApplicationTest {
     }
 
     private GetDimens getDimens(Size size) {
-        TargetUi targetUi = new TargetUi(activityRule.getActivity());
-        GetPath getPath = new GetPath(targetUi);
         Config config = new Config();
         config.setSize(size);
+        TargetUi targetUi = new TargetUi(activityRule.getActivity());
+        ImageUtils imageUtils = new ImageUtils(targetUi, config);
+        DownloadImage downloadImage = new DownloadImage(targetUi, imageUtils);
+        GetPath getPath = new GetPath(targetUi, imageUtils, downloadImage);
         return new GetDimens(targetUi, config, getPath);
     }
 

--- a/app/src/main/java/com/fuck_boilerplate/rx_paparazzo/sample/activities/SampleActivity.java
+++ b/app/src/main/java/com/fuck_boilerplate/rx_paparazzo/sample/activities/SampleActivity.java
@@ -79,6 +79,9 @@ public class SampleActivity extends AppCompatActivity implements Testable {
                     }
 
                     response.targetUI().loadImage(response.data());
+                }, throwable -> {
+                    throwable.printStackTrace();
+                    Toast.makeText(getApplicationContext(), "ERROR " + throwable.getMessage(), Toast.LENGTH_SHORT).show();
                 });
     }
 
@@ -99,6 +102,9 @@ public class SampleActivity extends AppCompatActivity implements Testable {
                     }
 
                     response.targetUI().loadImage(response.data());
+                }, throwable -> {
+                    throwable.printStackTrace();
+                    Toast.makeText(getApplicationContext(), "ERROR " + throwable.getMessage(), Toast.LENGTH_SHORT).show();
                 });
     }
 
@@ -120,6 +126,9 @@ public class SampleActivity extends AppCompatActivity implements Testable {
                     }
 
                     response.targetUI().loadImage(response.data());
+                }, throwable -> {
+                    throwable.printStackTrace();
+                    Toast.makeText(getApplicationContext(), "ERROR " + throwable.getMessage(), Toast.LENGTH_SHORT).show();
                 });
     }
 
@@ -138,6 +147,9 @@ public class SampleActivity extends AppCompatActivity implements Testable {
 
                     if (response.data().size() == 1) response.targetUI().loadImage(response.data().get(0));
                     else response.targetUI().loadImages(response.data());
+                }, throwable -> {
+                    throwable.printStackTrace();
+                    Toast.makeText(getApplicationContext(), "ERROR " + throwable.getMessage(), Toast.LENGTH_SHORT).show();
                 });
     }
 

--- a/rx_paparazzo/src/main/AndroidManifest.xml
+++ b/rx_paparazzo/src/main/AndroidManifest.xml
@@ -4,6 +4,7 @@
     <uses-permission android:name="android.permission.READ_EXTERNAL_STORAGE"/>
     <uses-permission android:name="android.permission.WRITE_EXTERNAL_STORAGE"/>
     <uses-permission android:name="android.permission.CAMERA"/>
+    <uses-permission android:name="android.permission.INTERNET"/>
 
     <application
         android:allowBackup="true"

--- a/rx_paparazzo/src/main/java/com/fuck_boilerplate/rx_paparazzo/interactors/DownloadImage.java
+++ b/rx_paparazzo/src/main/java/com/fuck_boilerplate/rx_paparazzo/interactors/DownloadImage.java
@@ -1,0 +1,97 @@
+/*
+ * Copyright 2016 FuckBoilerplate
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.fuck_boilerplate.rx_paparazzo.interactors;
+
+import android.net.Uri;
+
+import com.fuck_boilerplate.rx_paparazzo.entities.TargetUi;
+import com.fuck_boilerplate.rx_paparazzo.entities.UserCanceledException;
+
+import java.io.BufferedInputStream;
+import java.io.File;
+import java.io.InputStream;
+import java.net.URL;
+import java.net.URLConnection;
+import java.util.concurrent.TimeUnit;
+
+import rx.Observable;
+import rx.Subscriber;
+import rx.android.schedulers.AndroidSchedulers;
+import rx.exceptions.Exceptions;
+import rx.functions.Func0;
+import rx.schedulers.Schedulers;
+
+public final class DownloadImage extends UseCase<String> {
+    private final TargetUi targetUi;
+    private final ImageUtils imageUtils;
+    private Uri uri;
+
+    public DownloadImage(TargetUi targetUi, ImageUtils imageUtils) {
+        this.targetUi = targetUi;
+        this.imageUtils = imageUtils;
+    }
+
+    @Override Observable<String> react() {
+        return getObservableDownloadFile();
+    }
+
+    public DownloadImage with(Uri uri) {
+        this.uri = uri;
+        return this;
+    }
+
+    private Observable<String> getObservableDownloadFile() {
+        return Observable.create(new Observable.OnSubscribe<String>() {
+            @Override
+            public void call(Subscriber<? super String> subscriber) {
+                try {
+                    if ("content".equalsIgnoreCase(uri.getScheme())){
+                        subscriber.onNext(downloadContent());
+                    }
+                    else {
+                        subscriber.onNext(downloadFile());
+                    }
+
+                    subscriber.onCompleted();
+                } catch (Exception e) {
+                    subscriber.onNext(null);
+                    subscriber.onError(e);
+                }
+            }
+        })
+        .subscribeOn(Schedulers.io())
+        .observeOn(AndroidSchedulers.mainThread());
+    }
+
+    private String downloadFile() throws Exception {
+        URL url = new URL(uri.toString());
+        URLConnection connection = url.openConnection();
+        connection.connect();
+        InputStream inputStream = new BufferedInputStream(url.openStream(), 1024);
+        File file = imageUtils.getOutputFile();
+
+        imageUtils.copy(inputStream, file);
+        return file.getAbsolutePath();
+    }
+
+    private String downloadContent() throws Exception {
+        InputStream inputStream = targetUi.getContext().getContentResolver().openInputStream(uri);
+        File file = imageUtils.getOutputFile();
+        imageUtils.copy(inputStream, file);
+        return file.getAbsolutePath();
+    }
+}

--- a/rx_paparazzo/src/main/java/com/fuck_boilerplate/rx_paparazzo/interactors/GetPath.java
+++ b/rx_paparazzo/src/main/java/com/fuck_boilerplate/rx_paparazzo/interactors/GetPath.java
@@ -28,15 +28,24 @@ import android.provider.MediaStore;
 
 import com.fuck_boilerplate.rx_paparazzo.entities.TargetUi;
 
+import java.io.File;
+import java.io.InputStream;
+
 import rx.Observable;
+import rx.android.schedulers.AndroidSchedulers;
 import rx.exceptions.Exceptions;
+import rx.schedulers.Schedulers;
 
 public final class GetPath extends UseCase<String> {
     private final TargetUi targetUi;
+    private final ImageUtils imageUtils;
+    private final DownloadImage downloadImage;
     private Uri uri;
 
-     public GetPath(TargetUi targetUi) {
+    public GetPath(TargetUi targetUi, ImageUtils imageUtils, DownloadImage downloadImage) {
         this.targetUi = targetUi;
+        this.imageUtils = imageUtils;
+        this.downloadImage = downloadImage;
     }
 
     public GetPath with(Uri uri) {
@@ -45,13 +54,14 @@ public final class GetPath extends UseCase<String> {
     }
 
     @Override public Observable<String> react() {
-        return Observable.just(getPath());
+        return getPath();
     }
 
     @SuppressLint("NewApi")
-    private String getPath() {
+    private Observable<String> getPath() {
         boolean isKitKat = Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT;
         Context context = targetUi.activity();
+        String filePath = null;
 
         if (uri == null) {
             return null;
@@ -61,13 +71,12 @@ public final class GetPath extends UseCase<String> {
             if (isExternalStorageDocument(uri)) {
                 Document document = getDocument(uri);
                 if ("primary".equalsIgnoreCase(document.type)) {
-                    return Environment.getExternalStorageDirectory() + "/" + document.id;
+                    filePath = Environment.getExternalStorageDirectory() + "/" + document.id;
                 }
-                return null;
             } else if (isDownloadsDocument(uri)) {
                 String id = DocumentsContract.getDocumentId(uri);
                 Uri contentUri = ContentUris.withAppendedId( Uri.parse("content://downloads/public_downloads"), Long.valueOf(id));
-                return getDataColumn(context, contentUri, null, null);
+                filePath = getDataColumn(context, contentUri, null, null);
             } else if (isMediaDocument(uri)) {
                 Document document = getDocument(uri);
                 Uri contentUri = null;
@@ -81,17 +90,21 @@ public final class GetPath extends UseCase<String> {
                     contentUri = MediaStore.Audio.Media.EXTERNAL_CONTENT_URI;
                 }
 
-                return getDataColumn(context, contentUri, "_id=?", new String[] {document.id});
+                filePath = getDataColumn(context, contentUri, "_id=?", new String[] {document.id});
             }
         }
         else if ("content".equalsIgnoreCase(uri.getScheme())) {
-            return getDataColumn(context, uri, null, null);
+            filePath = getDataColumn(context, uri, null, null);
         }
         else if ("file".equalsIgnoreCase(uri.getScheme())) {
-            return uri.getPath();
+            filePath = uri.getPath();
         }
 
-        return null;
+        if (filePath == null) {
+            return downloadImage.with(uri).react();
+        }
+
+        return Observable.just(filePath);
     }
 
     private class Document {

--- a/rx_paparazzo/src/main/java/com/fuck_boilerplate/rx_paparazzo/interactors/ImageUtils.java
+++ b/rx_paparazzo/src/main/java/com/fuck_boilerplate/rx_paparazzo/interactors/ImageUtils.java
@@ -1,0 +1,286 @@
+/*
+ * Copyright 2016 FuckBoilerplate
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.fuck_boilerplate.rx_paparazzo.interactors;
+
+
+import android.content.Context;
+import android.graphics.Bitmap;
+import android.graphics.BitmapFactory;
+import android.graphics.Matrix;
+import android.media.ExifInterface;
+import android.os.Environment;
+import android.text.TextUtils;
+
+import com.fuck_boilerplate.rx_paparazzo.entities.Config;
+import com.fuck_boilerplate.rx_paparazzo.entities.Size;
+import com.fuck_boilerplate.rx_paparazzo.entities.TargetUi;
+
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.text.SimpleDateFormat;
+import java.util.Date;
+import java.util.Locale;
+
+import rx.exceptions.Exceptions;
+
+public final class ImageUtils {
+    private static final String DATE_FORMAT = "ddMMyyyy_HHmmss_SSS";
+    private static final String LOCALE_EN = "en";
+    private final TargetUi targetUi;
+    private final Config config;
+
+    public ImageUtils(TargetUi targetUi, Config config) {
+        this.targetUi = targetUi;
+        this.config = config;
+    }
+
+    public File getOutputFile() {
+        String filename = new SimpleDateFormat(DATE_FORMAT, new Locale(LOCALE_EN)).format(new Date());
+        filename = "IMG-" + filename + ".jpg";
+        String dirname = getApplicationName(targetUi.getContext());
+        File dir = getPublicDir(null, dirname);
+        return new File(dir.getAbsolutePath(), filename);
+    }
+
+    private String getApplicationName(Context context) {
+        int stringId = context.getApplicationInfo().labelRes;
+        return context.getString(stringId);
+    }
+
+    private File getPublicDir(String dirRoot, String dirname) {
+        File storageDir = null;
+
+        if (!config.useInternalStorage()) {
+            if (Environment.getExternalStorageState().equals(Environment.MEDIA_MOUNTED)) {
+                File dir = (dirRoot != null) ? Environment.getExternalStoragePublicDirectory(dirRoot) : Environment.getExternalStorageDirectory();
+                storageDir = new File(dir, dirname);
+
+                if (!storageDir.exists() && !storageDir.mkdirs()) {
+                    storageDir = null;
+                }
+            }
+        }
+
+        if (storageDir == null) {
+            storageDir = getPrivateDir(dirname);
+        }
+
+        return storageDir;
+    }
+
+    private File getPrivateDir(String dirname) {
+        File dir = targetUi.getContext().getFilesDir();
+        File storageDir = TextUtils.isEmpty(dirname) ? dir : new File(dir, dirname);
+
+        // Create the storage directory if it does not exist
+        if (!storageDir.exists() && !storageDir.mkdirs()) {
+            storageDir = null;
+        }
+        return storageDir;
+    }
+
+    public String scaleImage(String filePath, String filePathOutput, int[] dimens) {
+        if (config.getSize() == Size.Original) {
+            return copyFile(filePath, filePathOutput);
+        }
+
+        boolean rotateIfNeeded = false;
+        Bitmap bitmap = handleSamplingAndRotationBitmap(filePath, dimens[0], dimens[1], rotateIfNeeded);
+
+        if (bitmap == null) {
+            return copyFile(filePath, filePathOutput);
+        }
+
+        File file = new File(filePath);
+        File fileScaled = new File(filePathOutput);
+        bitmap2file(bitmap, fileScaled, Bitmap.CompressFormat.JPEG);
+        copyExifRotation(file, fileScaled, rotateIfNeeded);
+
+        return filePathOutput;
+    }
+
+    private String copyFile(String filePath, String newfilePath) {
+        File file = new File(filePath);
+        File fileOutput = new File(newfilePath);
+
+        copy(file, fileOutput);
+        return newfilePath;
+    }
+
+    public void copy(InputStream in, File dst) {
+        OutputStream out = null;
+
+        try {
+            out = new FileOutputStream(dst);
+            byte[] buffer = new byte[1024];
+            int length;
+
+            while ((length = in.read(buffer)) > 0) {
+                out.write(buffer, 0, length);
+            }
+        } catch (IOException e) {
+            e.printStackTrace();
+            throw Exceptions.propagate(e);
+        } finally {
+            try {
+                if (in != null) {
+                    in.close();
+                }
+                if (out != null) {
+                    out.flush();
+                    out.close();
+                }
+            } catch (IOException e) {
+                // nothing
+            }
+        }
+    }
+
+    private void copy(File src, File dst) {
+        try {
+            InputStream in = new FileInputStream(src);
+            copy(in, dst);
+        } catch (IOException e) {
+            e.printStackTrace();
+            throw Exceptions.propagate(e);
+        }
+    }
+
+    private boolean bitmap2file(Bitmap bitmap, File file, Bitmap.CompressFormat compressFormat) {
+        FileOutputStream fileOutputStream = null;
+        try {
+            fileOutputStream = new FileOutputStream(file);
+            bitmap.compress(compressFormat, 90, fileOutputStream);
+            return true;
+        } catch (Exception e) {
+            e.printStackTrace();
+            throw Exceptions.propagate(e);
+        } finally {
+            try {
+                if (fileOutputStream != null) {
+                    fileOutputStream.flush();
+                    fileOutputStream.close();
+                }
+            } catch (IOException e) {
+                // nothing
+            }
+        }
+    }
+
+    private boolean copyExifRotation(File sourceFile, File destFile, boolean rotateIfNeeded) {
+        if (sourceFile == null || destFile == null) {
+            return false;
+        }
+        try {
+            ExifInterface exifSource = new ExifInterface(sourceFile.getAbsolutePath());
+            ExifInterface exifDest = new ExifInterface(destFile.getAbsolutePath());
+            String value = rotateIfNeeded ? String.valueOf(ExifInterface.ORIENTATION_NORMAL) : exifSource.getAttribute(ExifInterface.TAG_ORIENTATION);
+            exifDest.setAttribute(ExifInterface.TAG_ORIENTATION, value);
+            exifDest.saveAttributes();
+            return true;
+        } catch (IOException e) {
+            e.printStackTrace();
+            throw Exceptions.propagate(e);
+        }
+    }
+
+    private Bitmap handleSamplingAndRotationBitmap(String filePath, int maxWidth, int maxHeight, boolean rotateIfNeeded) {
+        BitmapFactory.Options options = new BitmapFactory.Options();
+        options.inJustDecodeBounds = true;
+        BitmapFactory.decodeFile(filePath, options);
+        options.inSampleSize = (maxWidth <= 0 || maxHeight <= 0) ? 1 : calculateInSampleSize(options, maxWidth, maxHeight);
+
+        if (options.inSampleSize == 1)
+            return null;
+
+        options.inJustDecodeBounds = false;
+        Bitmap bitmap = BitmapFactory.decodeFile(filePath, options);
+
+        if (rotateIfNeeded)
+            bitmap = rotateImageIfRequired(bitmap, filePath);
+
+        return bitmap;
+    }
+
+    private int calculateInSampleSize(BitmapFactory.Options options, int maxWidth, int maxHeight) {
+        int inSampleSize = 1;
+        int[] dimensPortrait = getDimensionsPortrait(options.outWidth, options.outHeight);
+        int[] maxDimensPortrait = getDimensionsPortrait(maxWidth, maxHeight);
+        float width = dimensPortrait[0];
+        float height = dimensPortrait[1];
+        float newMaxWidth = maxDimensPortrait[0];
+        float newMaxHeight = maxDimensPortrait[1];
+
+        if (height > newMaxHeight || width > newMaxWidth) {
+            int heightRatio = Math.round(height / newMaxHeight);
+            int widthRatio = Math.round(width / newMaxWidth);
+            inSampleSize = heightRatio < widthRatio ? heightRatio : widthRatio;
+            float totalPixels = width * height;
+            float totalReqPixels = newMaxWidth * newMaxHeight * 2;
+
+            while (totalPixels / (inSampleSize * inSampleSize) > totalReqPixels) {
+                inSampleSize++;
+            }
+        }
+
+        return inSampleSize;
+    }
+
+    private int[] getDimensionsPortrait(int width, int height) {
+        if (width < height) {
+            return new int[]{width, height};
+        } else {
+            return new int[]{height, width};
+        }
+    }
+
+    private Bitmap rotateImageIfRequired(Bitmap img, String filePath) {
+        try {
+            ExifInterface ei = new ExifInterface(filePath);
+            int orientation = ei.getAttributeInt(ExifInterface.TAG_ORIENTATION, ExifInterface.ORIENTATION_UNDEFINED);
+            Bitmap processedImage = img;
+            switch (orientation) {
+                case ExifInterface.ORIENTATION_ROTATE_90:
+                    processedImage = rotateImage(img, 90);
+                    break;
+                case ExifInterface.ORIENTATION_ROTATE_180:
+                    processedImage = rotateImage(img, 180);
+                    break;
+                case ExifInterface.ORIENTATION_ROTATE_270:
+                    processedImage = rotateImage(img, 270);
+                    break;
+                default:
+                    break;
+            }
+            return processedImage;
+        } catch (IOException e) {
+            throw Exceptions.propagate(e);
+        }
+    }
+
+    private Bitmap rotateImage(Bitmap img, int degree) {
+        Matrix matrix = new Matrix();
+        matrix.postRotate(degree);
+        Bitmap rotatedImg = Bitmap.createBitmap(img, 0, 0, img.getWidth(), img.getHeight(), matrix, true);
+        img.recycle();
+        return rotatedImg;
+    }
+}

--- a/rx_paparazzo/src/main/java/com/fuck_boilerplate/rx_paparazzo/interactors/SaveImage.java
+++ b/rx_paparazzo/src/main/java/com/fuck_boilerplate/rx_paparazzo/interactors/SaveImage.java
@@ -16,49 +16,30 @@
 
 package com.fuck_boilerplate.rx_paparazzo.interactors;
 
-import android.content.Context;
-import android.graphics.Bitmap;
-import android.graphics.BitmapFactory;
-import android.graphics.Matrix;
-import android.media.ExifInterface;
 import android.media.MediaScannerConnection;
 import android.net.Uri;
-import android.os.Environment;
-import android.text.TextUtils;
 
 import com.fuck_boilerplate.rx_paparazzo.entities.Config;
-import com.fuck_boilerplate.rx_paparazzo.entities.Size;
 import com.fuck_boilerplate.rx_paparazzo.entities.TargetUi;
 
-import java.io.File;
-import java.io.FileInputStream;
-import java.io.FileOutputStream;
-import java.io.IOException;
-import java.io.InputStream;
-import java.io.OutputStream;
-import java.text.SimpleDateFormat;
-import java.util.Date;
-import java.util.Locale;
-
 import rx.Observable;
-import rx.exceptions.Exceptions;
 import rx.functions.Func1;
 import rx.functions.Func3;
 
 public final class SaveImage extends UseCase<String> {
-    private static final String DATE_FORMAT = "ddMMyyyy_HHmmss_SSS";
-    private static final String LOCALE_EN = "en";
     private final TargetUi targetUi;
     private final Config config;
     private final GetPath getPath;
     private final GetDimens getDimens;
+    private final ImageUtils imageUtils;
     private Uri uri;
 
-     public SaveImage(TargetUi targetUi, Config config, GetPath getPath, GetDimens getDimens) {
+    public SaveImage(TargetUi targetUi, Config config, GetPath getPath, GetDimens getDimens, ImageUtils imageUtils) {
         this.targetUi = targetUi;
         this.config = config;
         this.getPath = getPath;
         this.getDimens = getDimens;
+        this.imageUtils = imageUtils;
     }
 
     public SaveImage with(Uri uri) {
@@ -66,7 +47,8 @@ public final class SaveImage extends UseCase<String> {
         return this;
     }
 
-    @Override public Observable<String> react() {
+    @Override
+    public Observable<String> react() {
         return getOutputUri()
                 .flatMap(new Func1<Uri, Observable<String>>() {
                     @Override
@@ -75,230 +57,18 @@ public final class SaveImage extends UseCase<String> {
                                 getPath.with(outputUri).react(), getDimens.with(uri).react(),
                                 new Func3<String, String, int[], String>() {
                                     @Override
-                                    public String call(String filePath, String filePathOutput,
-                                            int[] dimens) {
-                                        String filepath = SaveImage.this.scaleImage(filePath,
-                                                filePathOutput, dimens);
-                                        MediaScannerConnection.scanFile(targetUi.getContext(),
-                                                new String[] {
-                                                        filepath
-                                        }, new String[] {
-                                                "image/jpeg"
-                                        }, null);
-                                        return filepath;
+                                    public String call(String filePath, String filePathOutput, int[] dimens) {
+                                        String filePathScaled = imageUtils.scaleImage(filePath, filePathOutput, dimens);
+                                        MediaScannerConnection.scanFile(targetUi.getContext(), new String[]{filePathScaled}, new String[]{"image/jpeg"}, null);
+
+                                        return filePathScaled;
                                     }
                                 });
                     }
-        });
+                });
     }
 
     private Observable<Uri> getOutputUri() {
-        return Observable.just(Uri.fromFile(getOutputFile()));
-    }
-
-    private File getOutputFile() {
-        String filename = new SimpleDateFormat(DATE_FORMAT, new Locale(LOCALE_EN)).format(new Date());
-        filename = "IMG-" + filename + ".jpg";
-        String dirname = getApplicationName(targetUi.getContext());
-        File dir = getPublicDir(null, dirname);
-        return new File(dir.getAbsolutePath(), filename);
-    }
-
-    private String getApplicationName(Context context) {
-        int stringId = context.getApplicationInfo().labelRes;
-        return context.getString(stringId);
-    }
-
-    private File getPublicDir(String dirRoot, String dirname) {
-        File storageDir = null;
-
-        if (!config.useInternalStorage()) {
-            if (Environment.getExternalStorageState().equals(Environment.MEDIA_MOUNTED)) {
-                File dir = (dirRoot != null) ? Environment.getExternalStoragePublicDirectory(dirRoot) : Environment.getExternalStorageDirectory();
-                storageDir = new File(dir, dirname);
-
-                if (!storageDir.exists() && !storageDir.mkdirs()) {
-                    storageDir = null;
-                }
-            }
-        }
-
-        if (storageDir == null) {
-            storageDir = getPrivateDir(dirname);
-        }
-
-        return storageDir;
-    }
-
-    private File getPrivateDir(String dirname) {
-        File dir = targetUi.getContext().getFilesDir();
-        File storageDir = TextUtils.isEmpty(dirname) ? dir : new File(dir, dirname);
-
-        // Create the storage directory if it does not exist
-        if (!storageDir.exists() && !storageDir.mkdirs()) {
-            storageDir = null;
-        }
-        return storageDir;
-    }
-
-    private String scaleImage(String filePath, String filePathOutput, int[] dimens) {
-        if (config.getSize() == Size.Original) {
-            return copyFile(filePath, filePathOutput);
-        }
-
-        boolean rotateIfNeeded = false;
-        Bitmap bitmap = handleSamplingAndRotationBitmap(filePath, dimens[0], dimens[1], rotateIfNeeded);
-
-        if (bitmap == null) {
-            return copyFile(filePath, filePathOutput);
-        }
-
-        File file = new File(filePath);
-        File fileScaled = new File(filePathOutput);
-        bitmap2file(bitmap, fileScaled, Bitmap.CompressFormat.JPEG);
-        copyExifRotation(file, fileScaled, rotateIfNeeded);
-
-        return filePathOutput;
-    }
-
-    private String copyFile(String filePath, String newfilePath) {
-        File file = new File(filePath);
-        File fileOutput = new File(newfilePath);
-
-        copy(file, fileOutput);
-
-        return newfilePath;
-    }
-
-    private void copy(File src, File dst) {
-        try {
-            InputStream in = new FileInputStream(src);
-            OutputStream out = new FileOutputStream(dst);
-            byte[] buffer = new byte[1024];
-            int length;
-
-            while ((length = in.read(buffer)) > 0) {
-                out.write(buffer, 0, length);
-            }
-
-            in.close();
-            out.close();
-        } catch (IOException e) {
-            e.printStackTrace();
-            throw Exceptions.propagate(e);
-        }
-    }
-
-    private boolean bitmap2file(Bitmap bitmap, File file, Bitmap.CompressFormat compressFormat) {
-        try {
-            FileOutputStream fileOutputStream = new FileOutputStream(file);
-            bitmap.compress(compressFormat, 90, fileOutputStream);
-            fileOutputStream.flush();
-            fileOutputStream.close();
-            return true;
-        } catch (Exception e) {
-            e.printStackTrace();
-            throw Exceptions.propagate(e);
-        }
-    }
-
-    private boolean copyExifRotation(File sourceFile, File destFile, boolean rotateIfNeeded) {
-        if (sourceFile == null || destFile == null) {
-            return false;
-        }
-        try {
-            ExifInterface exifSource = new ExifInterface(sourceFile.getAbsolutePath());
-            ExifInterface exifDest = new ExifInterface(destFile.getAbsolutePath());
-            String value = rotateIfNeeded ? String.valueOf(ExifInterface.ORIENTATION_NORMAL) : exifSource.getAttribute(ExifInterface.TAG_ORIENTATION);
-            exifDest.setAttribute(ExifInterface.TAG_ORIENTATION, value);
-            exifDest.saveAttributes();
-            return true;
-        } catch (IOException e) {
-            e.printStackTrace();
-            throw Exceptions.propagate(e);
-        }
-    }
-
-    private Bitmap handleSamplingAndRotationBitmap(String filePath, int maxWidth, int maxHeight, boolean rotateIfNeeded) {
-        BitmapFactory.Options options = new BitmapFactory.Options();
-        options.inJustDecodeBounds = true;
-        BitmapFactory.decodeFile(filePath, options);
-        options.inSampleSize = (maxWidth <= 0 || maxHeight <= 0) ? 1 : calculateInSampleSize(options, maxWidth, maxHeight);
-
-        if (options.inSampleSize == 1)
-            return null;
-
-        options.inJustDecodeBounds = false;
-        Bitmap bitmap = BitmapFactory.decodeFile(filePath, options);
-
-        if (rotateIfNeeded)
-            bitmap = rotateImageIfRequired(bitmap, filePath);
-
-        return bitmap;
-    }
-
-    private int calculateInSampleSize(BitmapFactory.Options options, int maxWidth, int maxHeight) {
-        int inSampleSize = 1;
-        int[] dimensPortrait = getDimensionsPortrait(options.outWidth, options.outHeight);
-        int[] maxDimensPortrait = getDimensionsPortrait(maxWidth, maxHeight);
-        float width = dimensPortrait[0];
-        float height = dimensPortrait[1];
-        float newMaxWidth = maxDimensPortrait[0];
-        float newMaxHeight = maxDimensPortrait[1];
-
-        if (height > newMaxHeight || width > newMaxWidth) {
-            int heightRatio = Math.round(height / newMaxHeight);
-            int widthRatio = Math.round(width / newMaxWidth);
-            inSampleSize = heightRatio < widthRatio ? heightRatio : widthRatio;
-            float totalPixels = width * height;
-            float totalReqPixels = newMaxWidth * newMaxHeight * 2;
-
-            while (totalPixels / (inSampleSize * inSampleSize) > totalReqPixels) {
-                inSampleSize++;
-            }
-        }
-
-        return inSampleSize;
-    }
-
-    private int[] getDimensionsPortrait(int width, int height) {
-        if (width < height) {
-            return new int[]{width, height};
-        }
-        else {
-            return new int[]{height, width};
-        }
-    }
-
-    private Bitmap rotateImageIfRequired(Bitmap img, String filePath) {
-        try {
-            ExifInterface ei = new ExifInterface(filePath);
-            int orientation = ei.getAttributeInt(ExifInterface.TAG_ORIENTATION, ExifInterface.ORIENTATION_UNDEFINED);
-            Bitmap processedImage = img;
-            switch (orientation) {
-                case ExifInterface.ORIENTATION_ROTATE_90:
-                    processedImage = rotateImage(img, 90);
-                    break;
-                case ExifInterface.ORIENTATION_ROTATE_180:
-                    processedImage = rotateImage(img, 180);
-                    break;
-                case ExifInterface.ORIENTATION_ROTATE_270:
-                    processedImage = rotateImage(img, 270);
-                    break;
-                default:
-                    break;
-            }
-            return processedImage;
-        } catch (IOException e) {
-            throw Exceptions.propagate(e);
-        }
-    }
-
-    private Bitmap rotateImage(Bitmap img, int degree) {
-        Matrix matrix = new Matrix();
-        matrix.postRotate(degree);
-        Bitmap rotatedImg = Bitmap.createBitmap(img, 0, 0, img.getWidth(), img.getHeight(), matrix, true);
-        img.recycle();
-        return rotatedImg;
+        return Observable.just(Uri.fromFile(imageUtils.getOutputFile()));
     }
 }

--- a/rx_paparazzo/src/main/java/com/fuck_boilerplate/rx_paparazzo/internal/di/ApplicationComponentImpl.java
+++ b/rx_paparazzo/src/main/java/com/fuck_boilerplate/rx_paparazzo/internal/di/ApplicationComponentImpl.java
@@ -4,6 +4,8 @@ package com.fuck_boilerplate.rx_paparazzo.internal.di;
 import com.fuck_boilerplate.rx_paparazzo.entities.Config;
 import com.fuck_boilerplate.rx_paparazzo.entities.TargetUi;
 import com.fuck_boilerplate.rx_paparazzo.interactors.CropImage;
+import com.fuck_boilerplate.rx_paparazzo.interactors.DownloadImage;
+import com.fuck_boilerplate.rx_paparazzo.interactors.ImageUtils;
 import com.fuck_boilerplate.rx_paparazzo.interactors.GetDimens;
 import com.fuck_boilerplate.rx_paparazzo.interactors.GetPath;
 import com.fuck_boilerplate.rx_paparazzo.interactors.GrantPermissions;
@@ -16,6 +18,8 @@ import com.fuck_boilerplate.rx_paparazzo.workers.Camera;
 import com.fuck_boilerplate.rx_paparazzo.workers.Gallery;
 
 class ApplicationComponentImpl extends ApplicationComponent {
+    private final ImageUtils imageUtils;
+    private final DownloadImage downloadImage;
     private final StartIntent startIntent;
     private final GetPath getPath;
     private final GetDimens getDimens;
@@ -30,11 +34,13 @@ class ApplicationComponentImpl extends ApplicationComponent {
 
     public ApplicationComponentImpl(TargetUi ui, Config config) {
         startIntent = new StartIntent(ui);
-        getPath = new GetPath(ui);
+        imageUtils = new ImageUtils(ui, config);
+        downloadImage = new DownloadImage(ui, imageUtils);
+        getPath = new GetPath(ui, imageUtils, downloadImage);
         takePhoto = new TakePhoto(startIntent, ui);
         getDimens = new GetDimens(ui, config, getPath);
         cropImage = new CropImage(ui, config, startIntent, getPath, getDimens);
-        saveImage = new SaveImage(ui, config, getPath, getDimens);
+        saveImage = new SaveImage(ui, config, getPath, getDimens, imageUtils);
         grantPermissions = new GrantPermissions(ui);
         pickImages = new PickImages(startIntent);
         pickImage = new PickImage(startIntent, ui);

